src/tosyntax/FStar.ToSyntax.Resugar.fs

```diff
@@ -0,0 +1,614 @@
+          let (body, _) = args.Head in
+          begin match (SS.compress body).n with
+            | Tm_abs(xs, body, _) ->
+                let xs, body = SS.open_term xs body in
+                let xs = xs |> List.map (fun (b, qual) -> resugar_bv_as_binder b t.pos) in 
+                let pats, body = match (SS.compress body).n with
+                  | Tm_meta(e, m) ->
+                    let body = resugar_term e in
+                    let resugar_pats pats = List.map (fun es -> es |> List.map (fun (e, _) -> resugar_term e)) pats in
+                    let pats = match m with 
+                      | Meta_pattern pats -> resugar_pats pats
+                      | _ -> failwith "wrong pattern format for QForall" 
+                    in
+                    pats, body
+                  | _ -> [], resugar_term body
+                in
+                if op = "forall" then mk (A.QForall(xs, pats, body)) else mk (A.QExists(xs, pats, body))   
+                    
+            | _ -> failwith "wrong format to QForall"
+          end
+
+        | Some "alloc" ->
+          let (e, _ ) = args.Head in
+          resugar_term e
+
+        | Some op ->
+          let args = args |> List.map (fun (e, qual) ->
+            resugar_term e) in
+          mk (A.Op(op, args))
+      end
+    
+    | Tm_match(e, [(pat1, _, t1); (pat2, _, t2)]) when is_true_pat pat1 && is_wild_pat pat2 ->
+      mk (A.If(resugar_term e, resugar_term t1, resugar_term t2))
+
+    | Tm_match(e, branches) ->
+      let resugar_branch (pat, wopt,b) =
+        let pat = resugar_match_pat pat in
+        let wopt = match wopt with
+          | None -> None
+          | Some e -> Some (resugar_term e) in
+        let b = resugar_term b in
+        (pat, wopt, b) in
+        mk (A.Match(resugar_term e, List.map resugar_branch branches))
+
+    | Tm_ascribed(e, comp, _) ->
+      let term = match comp with
+          | Inl n -> (* term *)
+            resugar_term n
+          | Inr n -> (* comp *)
+            resugar_comp n in
+      mk (A.Ascribed(resugar_term e, term))
+
+    | Tm_let((is_rec, bnds), body) ->
+      let mk_pat a = A.mk_pattern a t.pos in
+      let resugar_one_binding bnd = 
+        let binders, term, is_pat_app = match (SS.compress bnd.lbdef).n with
+          | Tm_abs(b, t, _) -> 
+            let b, t = SS.open_term b t in
+            b, t, true
+          | _ -> [], bnd.lbdef, false
+        in 
+        let pat, term = match bnd.lbname with
+          | Inr fv -> mk_pat (A.PatName fv.fv_name.v), term
+          | Inl bv -> 
+            let x, term = SS.open_term [S.mk_binder bv] term in
+            let (bv, _) = x.Head in 
+            mk_pat (A.PatVar (bv_as_unique_ident bv, None)), term
+        in
+        if is_pat_app then
+          let args = binders |> List.map (fun (bv, _) -> mk_pat(A.PatVar (bv_as_unique_ident bv, None))) in
+          (mk_pat (A.PatApp (pat, args)), resugar_term term)
+        else
+          (pat, resugar_term term)
```

Comment:
Would calling "open_let_rec bnds body" take cares of opening the universe on both bnd.lbtyp and bnd.lbdef? Also, in resugar we only see terms, we don't have any reference to toplevel declarations.